common/variant/shredding.md

```diff
@@ -0,0 +1,244 @@
+
+```
+optional group variant_col {
+ optional binary untyped_value;
+ optional group object {
+  optional group a {
+   optional binary untyped_value;
+   optional int64 typed_value;
+  }
+  optional group b {
+   optional binary untyped_value;
+   optional group object {
+    optional group c {
+      optional binary untyped_value;
+      optional binary typed_value (STRING);
+    }
+   }
+  }
+ }
+}
+```
+
+| Variant Value | Top-level untyped_value | b.untyped_value | Non-null in a | Non-null in b.c |
+|---------------|--------------------------|---------------|---------------|
+| {a: 123, b: {c: “hello”}} | null | null | typed_value | typed_value |
+| {a: 1.23, b: {c: “123”}} | null | null | untyped_value | typed_value |
+| {a: [1,2,3], b: {c: null}} | null | null | untyped_value | untyped_value |
+| {a: 123, c: 456} | {c: 456} | null | typed_value | null |
+| {a: 123, b: {c: "hello", d: 456}} | null | {d: 456} | typed_value | typed_value |
+| [{a: 1, b: {c: 2}}, {a: 3, b: {c: 4}}] | [{a: 1, b: {c: 2}}, {a: 3, b: {c: 4}}] | null | null | null |
+
+# Parquet Layout
+
+The **array** and **object** fields represent Variant array and object types, respectively. Arrays must use the three-level list structure described in https://github.com/apache/parquet-format/blob/master/LogicalTypes.md.
+
+An **object** field must be a group. Each field name of this inner group corresponds to the Variant value's object field name. Each inner field's type is a recursively shredded variant value: that is, the fields of each object field must be one or more of **object**, **array**, **typed_value** or **untyped_value**.
+
+Similarly the elements of an **array** must be a group containing one or more of **object**, **array**, **typed_value** or **untyped_value**.
+
+Each leaf in the schema can store an arbitrary Variant value. It contains an **untyped_value** binary field and a **typed_value** field. If non-null, **untyped_value** represents the value stored as a Variant binary; the metadata and value of a normal Variant are concatenated. The **typed_value** field may be any type that has a corresponding Variant type. For each value in the data, at most one of the **typed_value** and **untyped_value** may be non-null. A writer may omit either field, which is equivalent to all rows being null.
+
+| typed_value | untyped_value | Meaning |
+|-------------|----------------|---------|
+| null | null | Field is missing in the reconstructed Variant. |
+| null | non-null | Field may be any type in the reconstructed Variant. |
+| non-null | null | Field has this column’s type in the reconstructed Variant. |
+| non-null | non-null | Invalid |
```

Comment:
What should implementations do when they encounter this case?

Reply:
Failing seems like the right choice.

An alternative might be to prefer `untyped_value` over the others, which could allow an engine to write some redundant data into the typed column if it knew how to interpret it later. E.g. storing "123" in a string column in addition to keeping it as an integer in the `untyped_value` column. This seems like added confusion for marginal benefit, though.